File: golang/q400/q400.go
package q400

import (
	"math"
)

func FindNthDigitPlanB(n int) int {
	c := 1
	r := 0
	for m := n; ; {
		rr := c * 9 * pow(10, c-1)
		m = m - rr
		if m <= 0 {
			break
		}
		c++
		r += rr
	}
	m := n - r - 1
	num := m/c + pow(10, c-1)
	index := c - m%c - 1
	return indexNum(num, index)
}

func indexNum(f, i int) int {
	nums := make([]int, 0)
	for f > 0 {
		nums = append(nums, f%10)
		f = f / 10
	}
	return nums[i]
}

func pow(x, y int) int {
	return int(math.Pow(float64(x), float64(y)))
}

File: hs/src/Q400/Q400.hs
module Q400.Q400
(
    findNthDigitPlanA,
    findNthDigitPlanB,
) where

findNthDigitPlanA :: Int -> Int
findNthDigitPlanA n | n <= 0 = 0
findNthDigitPlanA n = last . take n . concat . map (reverse . splitNum) $ [1..]

splitNum :: Int -> [Int]
splitNum 0 = []
splitNum x = (x `mod` 10): splitNum (x `div` 10);

findNthDigitPlanB :: Int -> Int
findNthDigitPlanB n | n <= 0 = 0
findNthDigitPlanB n = planB m c where
    c = findRange n
    m = n - (sum [ 9 * x * (pow 10 (x-1)) | x<-[0..(c-1)]])

findRange :: Int -> Int
findRange n = local n 1 where
        local x1 x2
            | x1 <= 0 = x2 -1
            | otherwise = local (x1 - x2 * 9 * (pow 10 (x2-1))) (x2+1)

planB :: Int -> Int -> Int
planB m c = (splitNum num)!!index where
        num = ((m-1) `div` c) + (pow 10 (c-1))
        index = c - ((m-1) `mod` c) - 1

pow :: Int -> Int -> Int
pow _ 0 = 1
pow x1 x2 = foldl (*) 1 . take x2 . repeat $ x1
